src/app/hooks/useRouteParams.ts

```diff
@@ -0,0 +1,31 @@
+import { useEffect, useState } from 'react'
+import { useParams } from 'react-router-dom'
+import { isValidOasisAddress, isValidEthAddress, getEvmBech32Address } from '../utils/helpers'
+
+const getOasisAddressFromParams = async (address: string): Promise<string> => {
+  if (isValidOasisAddress(address)) {
+    return address
+  } else if (isValidEthAddress(address)) {
+    return await getEvmBech32Address(address)
+  } else {
+    throw new Error('Invalid address')
+  }
+}
+
+export const useOasisAddressFromQueryParam = () => {
+  const [oasisAddress, setOasisAddress] = useState<string | undefined>()
+  const { address } = useParams()
+
+  useEffect(() => {
+    async function validateAddress() {
+      if (address) {
+        const oasisAddress = await getOasisAddressFromParams(address)
+        setOasisAddress(oasisAddress)
```

Comment:
> In case of eth address in url we need to derive oasis address to fetch emerald transactions or account details

API is supposed to support querying by eth address too

If we really really need this awkward async conversion anyway: I would move it into router, so it executes before initializing components.

Reply:
FWIW, the `async` is not because this requires a network roundtrip, it's just because hashing is async. This is the root: https://github.com/oasisprotocol/oasis-sdk/blob/b37e6da699df331f5a2ac62793f8be099c68469c/client-sdk/ts-web/core/src/address.ts#L14-L14
So it's not slow, but it might be ugly :)